src/src/com/microsoft/aad/adal/AuthenticationContext.java

```diff
@@ -1580,138 +1625,90 @@ private String getTokenHash(String token) {
-            if (userInfo != null) {
-                mKeyWithUserId = CacheKey.createCacheKey(request, userInfo.getUserId());
-                mKeyWithDisplayableId = CacheKey.createCacheKey(request, userInfo.getDisplayableId());
-            }
-        }
-    }
-
-    private RefreshItem getRefreshToken(final AuthenticationRequest request, boolean useMultiResourceRefreshToken) {
+    private RefreshItem getRefreshToken(final AuthenticationRequest request) {
         final String methodName = ":getRefreshToken";
         RefreshItem refreshItem = null;
         if (mTokenCacheStore != null) {
-            String userId = request.getUserId();
-            if (StringExtensions.IsNullOrBlank(userId)) {
-                // acquireTokenSilent expects userid field from UserInfo
-                userId = request.getLoginHint();
-            }
-            String keyUsed;
-            TokenCacheItem item;
-            if (!useMultiResourceRefreshToken) {
-                // target refreshToken for this resource only. CacheKey will
-                // include the resourceId in the cachekey
-                Logger.v(TAG, "Looking for regular refresh token");
-                keyUsed = CacheKey.createCacheKey(request, userId);
-                item = mTokenCacheStore.getItem(keyUsed);
-            } else {
-                // check multiResource item in cache. Cache key
-                // will not include resourceId in the cache key string.
-                Logger.v(TAG, "Looking for Multi Resource Refresh token");
-                keyUsed = CacheKey.createMultiResourceRefreshTokenKey(request, userId);
-                item = mTokenCacheStore.getItem(keyUsed);
-            }
+            Logger.v(TAG + methodName, "Look for resource specific refresh token in the cache.");
+            refreshItem = getRefreshTokenFromCache(KeyEntryType.REGULAR_REFRESH_TOKEN_ENTRY, request);
+            
+            // If regular RT is found, but it's not MRRT, shouldn't bother trying with FRT. All the token issued
+            // by EvoSts is MRRT. If regular RT is not MRRT, it basically means it not issued by EvoSts. FoCI feature
+            // is only visible to RT issued by evoSts. 
+            if (refreshItem == null || refreshItem.mMultiResource) {
+                final String regularRTLookupMsg = refreshItem == null ? "Resource specific RT does not exist, look for MRRT." 
+                        : "Found resource specific RT, which is also MRRT. Look for MRRT.";
+                Logger.v(TAG + methodName, regularRTLookupMsg);
+
+                final RefreshItem mrrtItem = getRefreshTokenFromCache(KeyEntryType.MULTI_RESOURCE_REFRESH_TOKEN_ENTRY, request);
+                refreshItem = mrrtItem;
+                
+                // If RT is found, but it's not marked as MRRT, shouldn't bother trying to find FRT. 
+                if (refreshItem == null || !StringExtensions.IsNullOrBlank(refreshItem.mFamilyClientId)) {
```

Comment:
Yeah, it appeared that currently we store RT with flag that it's not MRRT.
So there is no actually way to detect whether it was used RT or MRRT to get AT.
She is changing that for new tokens, but we should consider that old tokens don't have that